File: experiments-local/experiment-17/runs/agent=omp_language=python_model=openrouter/qwen/qwen3.7-plus_tooling=none/rep1/tools.py
from mcp.server.fastmcp import FastMCP
import pandas as pd
from data_loader import get_all_matches, load_fifa_data, normalize_team_name

mcp = FastMCP("BrazilianSoccer")

@mcp.tool()
def search_matches(
    team: str | None = None,
    opponent: str | None = None,
    season: int | None = None,
    competition: str | None = None,
    limit: int = 20
) -> str:
    """Search for matches based on team, opponent, season, or competition."""
    df = get_all_matches()
    
    if team:
        team_norm = normalize_team_name(team)
        mask = (df['home_team_norm'].str.contains(team_norm, na=False)) | \
               (df['away_team_norm'].str.contains(team_norm, na=False))
        df = df[mask]
        
    if opponent:
        opp_norm = normalize_team_name(opponent)
        mask = (df['home_team_norm'].str.contains(opp_norm, na=False)) | \
               (df['away_team_norm'].str.contains(opp_norm, na=False))
        df = df[mask]
        
    if season is not None:
        df = df[df['season'] == season]
        
    if competition:
        comp_norm = competition.lower()
        if "brasileir" in comp_norm:
            mask = df['competition'].str.lower().str.contains("brasileir", na=False) | \
                   df['competition'].str.lower().str.contains("serie a", na=False)
            df = df[mask]
        else:
            df = df[df['competition'].str.lower().str.contains(comp_norm, na=False)]
        
    df = df.sort_values('datetime', ascending=False).head(limit)
    
    results = []
    for _, row in df.iterrows():
        date_str = str(row['datetime'].date()) if pd.notna(row['datetime']) else "Unknown"
        h_goals = int(row['home_goal']) if pd.notna(row['home_goal']) else 0
        a_goals = int(row['away_goal']) if pd.notna(row['away_goal']) else 0
        season_str = str(int(row['season'])) if pd.notna(row['season']) else "Unknown"
        results.append(f"- {date_str}: {row['home_team']} {h_goals}-{a_goals} {row['away_team']} ({row['competition']}, Season {season_str})")
        
    if not results:
        return "No matches found matching the criteria."
        
    return "\n".join(results)

@mcp.tool()
def get_team_statistics(team: str, season: int | None = None, competition: str | None = None) -> str:
    """Get win/loss/draw records, goals scored/conceded, and win rate for a team."""
    df = get_all_matches()
    team_norm = normalize_team_name(team)
    
    if season is not None:
        df = df[df['season'] == season]
    if competition:
        comp_norm = competition.lower()
        if "brasileir" in comp_norm:
            mask = df['competition'].str.lower().str.contains("brasileir", na=False) | \
                   df['competition'].str.lower().str.contains("serie a", na=False)
            df = df[mask]
        else:
            df = df[df['competition'].str.lower().str.contains(comp_norm, na=False)]
        
    home_mask = df['home_team_norm'].str.contains(team_norm, na=False)
    away_mask = df['away_team_norm'].str.contains(team_norm, na=False)
    team_matches = df[home_mask | away_mask]
    
    if team_matches.empty:
        return f"No match data found for {team}."
        
    wins = 0
    draws = 0
    losses = 0
    goals_for = 0
    goals_against = 0
    
    for _, row in team_matches.iterrows():
        is_home = home_mask.loc[row.name]
        team_goals = int(row['home_goal']) if pd.notna(row['home_goal']) and is_home else int(row['away_goal']) if pd.notna(row['away_goal']) else 0
        opp_goals = int(row['away_goal']) if pd.notna(row['away_goal']) and is_home else int(row['home_goal']) if pd.notna(row['home_goal']) else 0
        
        goals_for += team_goals
        goals_against += opp_goals
        
        if team_goals > opp_goals:
            wins += 1
        elif team_goals == opp_goals:
            draws += 1
        else:
            losses += 1
            
    total_matches = len(team_matches)
    win_rate = (wins / total_matches) * 100 if total_matches > 0 else 0
    
    comp_str = f" - {competition}" if competition else ""
    season_str = f" (Season {season})" if season is not None else ""
    
    return f"""{team} Statistics{season_str}{comp_str}:
- Matches: {total_matches}
- Wins: {wins}, Draws: {draws}, Losses: {losses}
- Goals For: {goals_for}, Goals Against: {goals_against}
- Win Rate: {win_rate:.1f}%"""

@mcp.tool()
def get_head_to_head(team1: str, team2: str, limit: int = 10) -> str:
    """Get head-to-head record and recent matches between two teams."""
    df = get_all_matches()
    t1_norm = normalize_team_name(team1)
    t2_norm = normalize_team_name(team2)
    
    mask = (
        (df['home_team_norm'].str.contains(t1_norm, na=False) & df['away_team_norm'].str.contains(t2_norm, na=False)) |
        (df['home_team_norm'].str.contains(t2_norm, na=False) & df['away_team_norm'].str.contains(t1_norm, na=False))
    )
    
    h2h_matches = df[mask].sort_values('datetime', ascending=False).head(limit)
    
    if h2h_matches.empty:
        return f"No head-to-head matches found between {team1} and {team2}."
        
    t1_wins = 0
    t2_wins = 0
    draws = 0
    
    results = []
    for _, row in h2h_matches.iterrows():
        h_norm = row['home_team_norm']
        a_norm = row['away_team_norm']
        h_goals = int(row['home_goal']) if pd.notna(row['home_goal']) else 0
        a_goals = int(row['away_goal']) if pd.notna(row['away_goal']) else 0
        
        if h_goals > a_goals:
            if t1_norm in h_norm: t1_wins += 1
            else: t2_wins += 1
        elif h_goals < a_goals:
            if t1_norm in a_norm: t1_wins += 1
            else: t2_wins += 1
        else:
            draws += 1
            
        date_str = str(row['datetime'].date()) if pd.notna(row['datetime']) else 'Unknown'
        results.append(f"- {date_str}: {row['home_team']} {h_goals}-{a_goals} {row['away_team']} ({row['competition']})")
        
    return f"""Head-to-Head: {team1} vs {team2}
Total matches in dataset: {len(h2h_matches)}
{team1} wins: {t1_wins}
{team2} wins: {t2_wins}
Draws: {draws}

Recent matches:
""" + "\n".join(results)

@mcp.tool()
def search_players(
    name: str | None = None,
    nationality: str | None = None,
    club: str | None = None,
    min_rating: int | None = None,
    position: str | None = None,
    limit: int = 10
) -> str:
    """Search for players in the FIFA database by name, nationality, club, rating, or position."""
    df = load_fifa_data()
    
    if name:
        name_norm = name.lower()
        df = df[df['Name'].str.lower().str.contains(name_norm, na=False)]
        
    if nationality:
        nat_norm = nationality.lower()
        df = df[df['Nationality'].str.lower().str.contains(nat_norm, na=False)]
        
    if club:
        club_norm = normalize_team_name(club)
        df = df[df['Club'].apply(normalize_team_name).str.contains(club_norm, na=False)]
        
    if min_rating is not None:
        df = df[df['Overall'] >= min_rating]
        
    if position:
        pos_norm = position.upper()
        df = df[df['Position'].str.contains(pos_norm, na=False)]
        
    df = df.sort_values('Overall', ascending=False).head(limit)
    
    if df.empty:
        return "No players found matching the criteria."
        
    results = []
    for _, row in df.iterrows():
        results.append(f"- {row['Name']} (Overall: {int(row['Overall'])}, Position: {row['Position']}, Club: {row['Club']}, Nationality: {row['Nationality']})")
        
    return "Top matching players:\n" + "\n".join(results)

@mcp.tool()
def get_competition_standings(competition: str, season: int) -> str:
    """Calculate and return the standings for a specific competition and season."""
    df = get_all_matches()
    comp_norm = competition.lower()
    # Map common names
    if "brasileir" in comp_norm:
        mask = df['competition'].str.lower().str.contains("brasileir", na=False) | \
               df['competition'].str.lower().str.contains("serie a", na=False)
        df = df[mask]
    else:
        df = df[df['competition'].str.lower().str.contains(comp_norm, na=False)]
    df = df[df['season'] == season]
    
    if df.empty:
        return f"No match data found for {competition} in {season}."
        
    standings = {}
    for _, row in df.iterrows():
        h_team = row['home_team_norm'].title()
        a_team = row['away_team_norm'].title()
        h_goals = int(row['home_goal']) if pd.notna(row['home_goal']) else 0
        a_goals = int(row['away_goal']) if pd.notna(row['away_goal']) else 0
        
        for team in [h_team, a_team]:
            if team not in standings:
                standings[team] = {'played': 0, 'won': 0, 'drawn': 0, 'lost': 0, 'gf': 0, 'ga': 0, 'pts': 0}
                
        standings[h_team]['played'] += 1
        standings[a_team]['played'] += 1
        standings[h_team]['gf'] += h_goals
        standings[h_team]['ga'] += a_goals
        standings[a_team]['gf'] += a_goals
        standings[a_team]['ga'] += h_goals
        
        if h_goals > a_goals:
            standings[h_team]['won'] += 1
            standings[h_team]['pts'] += 3
            standings[a_team]['lost'] += 1
        elif h_goals < a_goals:
            standings[a_team]['won'] += 1
            standings[a_team]['pts'] += 3
            standings[h_team]['lost'] += 1
        else:
            standings[h_team]['drawn'] += 1
            standings[h_team]['pts'] += 1
            standings[a_team]['drawn'] += 1
            standings[a_team]['pts'] += 1
            
    sorted_teams = sorted(
        standings.items(),
        key=lambda x: (x[1]['pts'], x[1]['gf'] - x[1]['ga'], x[1]['gf']),
        reverse=True
    )
    
    results = [f"{competition} {season} Standings:"]
    for i, (team, stats) in enumerate(sorted_teams, 1):
        gd = stats['gf'] - stats['ga']
        gd_str = f"+{gd}" if gd > 0 else str(gd)
        results.append(f"{i}. {team} - {stats['pts']} pts ({stats['won']}W, {stats['drawn']}D, {stats['lost']}L, GF: {stats['gf']}, GA: {stats['ga']}, GD: {gd_str})")
        
    return "\n".join(results)

@mcp.tool()
def get_statistical_analysis(metric: str, competition: str | None = None, season: int | None = None) -> str:
    """Get aggregated statistics like average goals, biggest wins, or home win rate."""
    df = get_all_matches()
    
    if competition:
        comp_norm = competition.lower()
        df = df[df['competition'].str.lower().str.contains(comp_norm, na=False)]
    if season is not None:
        df = df[df['season'] == season]
        
    if df.empty:
        return "No match data found for the specified criteria."
        
    metric = metric.lower()
    
    if "average goals" in metric or "goals per match" in metric:
        total_goals = (df['home_goal'].fillna(0) + df['away_goal'].fillna(0)).sum()
        avg_goals = total_goals / len(df)
        return f"Average goals per match: {avg_goals:.2f} (Total matches: {len(df)})"
        
    elif "home win" in metric or "home record" in metric:
        home_wins = (df['home_goal'] > df['away_goal']).sum()
        home_win_rate = (home_wins / len(df)) * 100
        return f"Home win rate: {home_win_rate:.1f}% ({home_wins} wins out of {len(df)} matches)"
        
    elif "biggest win" in metric or "biggest victory" in metric:
        df['goal_diff'] = (df['home_goal'].fillna(0) - df['away_goal'].fillna(0)).abs()
        biggest = df.sort_values('goal_diff', ascending=False).head(5)
        results = ["Biggest victories:"]
        for _, row in biggest.iterrows():
            diff = int(row['goal_diff'])
            date_str = str(row['datetime'].date()) if pd.notna(row['datetime']) else 'Unknown'
            h_goals = int(row['home_goal']) if pd.notna(row['home_goal']) else 0
            a_goals = int(row['away_goal']) if pd.notna(row['away_goal']) else 0
            results.append(f"- {date_str}: {row['home_team']} {h_goals}-{a_goals} {row['away_team']} ({row['competition']}, diff: {diff})")
        return "\n".join(results)
        
    return f"Metric '{metric}' not recognized. Try 'average goals', 'home win rate', or 'biggest win'."

File: experiments-local/experiment-17/runs/agent=omp_language=typescript_model=openrouter/qwen/qwen3.7-plus_tooling=none/rep3/src/tools.ts
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import { loadData } from "./dataLoader.js";
import { normalizeTeamName, parseDate } from "./utils.js";
import { Match, TeamStats, HeadToHead } from "./types.js";

export function registerTools(server: McpServer) {
  server.tool(
    "search_matches",
    "Search for matches by team, competition, season, or date range.",
    {
      team: z.string().optional().describe("Team name to search for (home or away)"),
      competition: z.string().optional().describe("Competition name (e.g., Brasileirão, Copa do Brasil, Libertadores)"),
      season: z.number().optional().describe("Year of the season"),
      startDate: z.string().optional().describe("Start date in YYYY-MM-DD format"),
      endDate: z.string().optional().describe("End date in YYYY-MM-DD format"),
      limit: z.number().optional().default(50).describe("Maximum number of results to return"),
    },
    async ({ team, competition, season, startDate, endDate, limit }) => {
      const data = loadData();
      const allMatches: Match[] = [
        ...data.brasileiraoMatches,
        ...data.cupMatches,
        ...data.libertadoresMatches,
        ...data.brFootballMatches,
        ...data.novoCampeonatoMatches,
      ];

      let results = allMatches;

      if (team) {
        const normalizedTeam = normalizeTeamName(team);
        results = results.filter((match) => {
          const home = normalizeTeamName("home_team" in match ? match.home_team : "home" in match ? match.home : "Equipe_mandante" in match ? match.Equipe_mandante : "");
          const away = normalizeTeamName("away_team" in match ? match.away_team : "away" in match ? match.away : "Equipe_visitante" in match ? match.Equipe_visitante : "");
          return home.includes(normalizedTeam) || away.includes(normalizedTeam);
        });
      }

      if (competition) {
        const normalizedComp = normalizeTeamName(competition);
        results = results.filter((match) => normalizeTeamName(match.competition).includes(normalizedComp));
      }

      if (season) {
        results = results.filter((match) => {
          const matchSeason = "season" in match ? match.season : "Ano" in match ? match.Ano : undefined;
          return matchSeason === season;
        });
      }

      if (startDate || endDate) {
        const start = startDate ? parseDate(startDate) : null;
        const end = endDate ? parseDate(endDate) : null;

        results = results.filter((match) => {
          const dateStr = "datetime" in match ? match.datetime : "date" in match ? match.date : "Data" in match ? match.Data : "";
          const matchDate = parseDate(dateStr);
          if (!matchDate) return false;
          if (start && matchDate < start) return false;
          if (end && matchDate > end) return false;
          return true;
        });
      }

      // Sort by date descending
      results.sort((a, b) => {
        const dateA = parseDate("datetime" in a ? a.datetime : "date" in a ? a.date : "Data" in a ? a.Data : "")?.getTime() || 0;
        const dateB = parseDate("datetime" in b ? b.datetime : "date" in b ? b.date : "Data" in b ? b.Data : "")?.getTime() || 0;
        return dateB - dateA;
      });

      const limited = results.slice(0, limit);

      return {
        content: [
          {
            type: "text",
            text: JSON.stringify({
              totalMatches: results.length,
              returnedMatches: limited.length,
              matches: limited.map((m) => {
                if ("home_team" in m) {
                  return {
                    date: m.datetime,
                    homeTeam: m.home_team,
                    awayTeam: m.away_team,
                    homeGoal: m.home_goal,
                    awayGoal: m.away_goal,
                    competition: m.competition,
                    season: "season" in m ? m.season : undefined,
                  };
                }
                if ("home" in m) {
                  return {
                    date: `${m.date} ${m.time}`,
                    homeTeam: m.home,
                    awayTeam: m.away,
                    homeGoal: m.home_goal,
                    awayGoal: m.away_goal,
                    competition: m.competition,
                  };
                }
                return {
                  date: m.Data,
                  homeTeam: m.Equipe_mandante,
                  awayTeam: m.Equipe_visitante,
                  homeGoal: m.Gols_mandante,
                  awayGoal: m.Gols_visitante,
                  competition: m.competition,
                  season: m.Ano,
                };
              }),
            }, null, 2),
          },
        ],
      };
    }
  );

  server.tool(
    "get_team_statistics",
    "Get win/loss/draw records and goals for/against for a specific team, optionally filtered by season or competition.",
    {
      team: z.string().describe("Team name to get statistics for"),
      season: z.number().optional().describe("Year of the season"),
      competition: z.string().optional().describe("Competition name"),
    },
    async ({ team, season, competition }) => {
      const data = loadData();
      const allMatches: Match[] = [
        ...data.brasileiraoMatches,
        ...data.cupMatches,
        ...data.libertadoresMatches,
        ...data.brFootballMatches,
        ...data.novoCampeonatoMatches,
      ];

      const normalizedTeam = normalizeTeamName(team);
      let matches = allMatches.filter((match) => {
        const home = normalizeTeamName("home_team" in match ? match.home_team : "home" in match ? match.home : "Equipe_mandante" in match ? match.Equipe_mandante : "");
        const away = normalizeTeamName("away_team" in match ? match.away_team : "away" in match ? match.away : "Equipe_visitante" in match ? match.Equipe_visitante : "");
        return home.includes(normalizedTeam) || away.includes(normalizedTeam);
      });

      if (season) {
        matches = matches.filter((match) => {
          const matchSeason = "season" in match ? match.season : "Ano" in match ? match.Ano : undefined;
          return matchSeason === season;
        });
      }

      if (competition) {
        const normalizedComp = normalizeTeamName(competition);
        matches = matches.filter((match) => normalizeTeamName(match.competition).includes(normalizedComp));
      }

      let wins = 0;
      let draws = 0;
      let losses = 0;
      let goalsFor = 0;
      let goalsAgainst = 0;

      for (const match of matches) {
        let isHome = false;
        let homeGoals = 0;
        let awayGoals = 0;

        if ("home_team" in match) {
          isHome = normalizeTeamName(match.home_team).includes(normalizedTeam);
          homeGoals = match.home_goal;
          awayGoals = match.away_goal;
        } else if ("home" in match) {
          isHome = normalizeTeamName(match.home).includes(normalizedTeam);
          homeGoals = match.home_goal;
          awayGoals = match.away_goal;
        } else {
          isHome = normalizeTeamName(match.Equipe_mandante).includes(normalizedTeam);
          homeGoals = match.Gols_mandante;
          awayGoals = match.Gols_visitante;
        }

        if (isHome) {
          goalsFor += homeGoals;
          goalsAgainst += awayGoals;
          if (homeGoals > awayGoals) wins++;
          else if (homeGoals === awayGoals) draws++;
          else losses++;
        } else {
          goalsFor += awayGoals;
          goalsAgainst += homeGoals;
          if (awayGoals > homeGoals) wins++;
          else if (awayGoals === homeGoals) draws++;
          else losses++;
        }
      }

      const winRate = matches.length > 0 ? ((wins / matches.length) * 100).toFixed(1) : "0.0";

      const stats: TeamStats = {
        team,
        season,
        competition,
        matches: matches.length,
        wins,
        draws,
        losses,
        goalsFor,
        goalsAgainst,
        winRate: parseFloat(winRate),
      };

      return {
        content: [
          {
            type: "text",
            text: JSON.stringify(stats, null, 2),
          },
        ],
      };
    }
  );

  server.tool(
    "search_players",
    "Search for players by name, nationality, club, or position, and filter by minimum overall rating.",
    {
      name: z.string().optional().describe("Player name or part of the name"),
      nationality: z.string().optional().describe("Player nationality (e.g., 'Brazil')"),
      club: z.string().optional().describe("Club name"),
      position: z.string().optional().describe("Playing position (e.g., 'ST', 'GK')"),
      minOverall: z.number().optional().describe("Minimum overall rating"),
      limit: z.number().optional().default(50).describe("Maximum number of results to return"),
    },
    async ({ name, nationality, club, position, minOverall, limit }) => {
      const data = loadData();
      let players = data.fifaPlayers;

      if (name) {
        const normalizedName = normalizeTeamName(name);
        players = players.filter((p) => normalizeTeamName(p.Name).includes(normalizedName));
      }

      if (nationality) {
        const normalizedNat = normalizeTeamName(nationality);
        players = players.filter((p) => normalizeTeamName(p.Nationality).includes(normalizedNat));
      }

      if (club) {
        const normalizedClub = normalizeTeamName(club);
        players = players.filter((p) => normalizeTeamName(p.Club).includes(normalizedClub));
      }

      if (position) {
        const normalizedPos = normalizeTeamName(position);
        players = players.filter((p) => normalizeTeamName(p.Position).includes(normalizedPos));
      }

      if (minOverall !== undefined) {
        players = players.filter((p) => p.Overall >= minOverall);
      }

      players.sort((a, b) => b.Overall - a.Overall);
      const limited = players.slice(0, limit);

      return {
        content: [
          {
            type: "text",
            text: JSON.stringify({
              totalPlayers: players.length,
              returnedPlayers: limited.length,
              players: limited.map((p) => ({
                name: p.Name,
                age: p.Age,
                nationality: p.Nationality,
                club: p.Club,
                position: p.Position,
                overall: p.Overall,
                potential: p.Potential,
              })),
            }, null, 2),
          },
        ],
      };
    }
  );

  server.tool(
    "get_head_to_head",
    "Get head-to-head match history and statistics between two specific teams.",
    {
      team1: z.string().describe("First team name"),
      team2: z.string().describe("Second team name"),
      limit: z.number().optional().default(20).describe("Maximum number of matches to return"),
    },
    async ({ team1, team2, limit }) => {
      const data = loadData();
      const allMatches: Match[] = [
        ...data.brasileiraoMatches,
        ...data.cupMatches,
        ...data.libertadoresMatches,
        ...data.brFootballMatches,
        ...data.novoCampeonatoMatches,
      ];

      const normTeam1 = normalizeTeamName(team1);
      const normTeam2 = normalizeTeamName(team2);

      const matches = allMatches.filter((match) => {
        let home = "";
        let away = "";
        if ("home_team" in match) {
          home = normalizeTeamName(match.home_team);
          away = normalizeTeamName(match.away_team);
        } else if ("home" in match) {
          home = normalizeTeamName(match.home);
          away = normalizeTeamName(match.away);
        } else {
          home = normalizeTeamName(match.Equipe_mandante);
          away = normalizeTeamName(match.Equipe_visitante);
        }

        const hasTeam1 = home.includes(normTeam1) || away.includes(normTeam1);
        const hasTeam2 = home.includes(normTeam2) || away.includes(normTeam2);
        return hasTeam1 && hasTeam2;
      });

      let team1Wins = 0;
      let team2Wins = 0;
      let draws = 0;

      for (const match of matches) {
        let homeGoals = 0;
        let awayGoals = 0;
        let homeTeamNorm = "";
        let awayTeamNorm = "";

        if ("home_team" in match) {
          homeGoals = match.home_goal;
          awayGoals = match.away_goal;
          homeTeamNorm = normalizeTeamName(match.home_team);
          awayTeamNorm = normalizeTeamName(match.away_team);
        } else if ("home" in match) {
          homeGoals = match.home_goal;
          awayGoals = match.away_goal;
          homeTeamNorm = normalizeTeamName(match.home);
          awayTeamNorm = normalizeTeamName(match.away);
        } else {
          homeGoals = match.Gols_mandante;
          awayGoals = match.Gols_visitante;
          homeTeamNorm = normalizeTeamName(match.Equipe_mandante);
          awayTeamNorm = normalizeTeamName(match.Equipe_visitante);
        }

        const isTeam1Home = homeTeamNorm.includes(normTeam1);
        if (homeGoals > awayGoals) {
          if (isTeam1Home) team1Wins++;
          else team2Wins++;
        } else if (awayGoals > homeGoals) {
          if (isTeam1Home) team2Wins++;
          else team1Wins++;
        } else {
          draws++;
        }
      }

      matches.sort((a, b) => {
        const dateA = parseDate("datetime" in a ? a.datetime : "date" in a ? a.date : "Data" in a ? a.Data : "")?.getTime() || 0;
        const dateB = parseDate("datetime" in b ? b.datetime : "date" in b ? b.date : "Data" in b ? b.Data : "")?.getTime() || 0;
        return dateB - dateA;
      });

      const limited = matches.slice(0, limit);

      const h2h: HeadToHead = {
        team1,
        team2,
        team1Wins,
        team2Wins,
        draws,
        matches: limited.map((m) => {
          if ("home_team" in m) {
            return { date: m.datetime, homeTeam: m.home_team, awayTeam: m.away_team, homeGoal: m.home_goal, awayGoal: m.away_goal, competition: m.competition };
          }
          if ("home" in m) {
            return { date: `${m.date} ${m.time}`, homeTeam: m.home, awayTeam: m.away, homeGoal: m.home_goal, awayGoal: m.away_goal, competition: m.competition };
          }
          return { date: m.Data, homeTeam: m.Equipe_mandante, awayTeam: m.Equipe_visitante, homeGoal: m.Gols_mandante, awayGoal: m.Gols_visitante, competition: m.competition, season: m.Ano };
        }),
      };

      return {
        content: [
          {
            type: "text",
            text: JSON.stringify(h2h, null, 2),
          },
        ],
      };
    }
  );
  server.tool(
    "get_competition_standings",
    "Calculate and return league standings for a specific competition and season based on match results.",
    {
      competition: z.string().describe("Competition name (e.g., 'Brasileirão', 'Copa do Brasil')"),
      season: z.number().describe("Year of the season"),
    },
    async ({ competition, season }) => {
      const data = loadData();
      const allMatches: Match[] = [
        ...data.brasileiraoMatches,
        ...data.cupMatches,
        ...data.libertadoresMatches,
        ...data.brFootballMatches,
        ...data.novoCampeonatoMatches,
      ];

      const normalizedComp = normalizeTeamName(competition);
      const matches = allMatches.filter((m) => {
        const matchSeason = "season" in m ? m.season : "Ano" in m ? m.Ano : undefined;
        return normalizeTeamName(m.competition).includes(normalizedComp) && matchSeason === season;
      });

      interface Standing {
        team: string;
        matches: number;
        wins: number;
        draws: number;
        losses: number;
        goalsFor: number;
        goalsAgainst: number;
        goalDifference: number;
        points: number;
      }

      const standingsMap = new Map<string, Standing>();

      const getOrCreate = (teamName: string): Standing => {
        if (!standingsMap.has(teamName)) {
          standingsMap.set(teamName, { team: teamName, matches: 0, wins: 0, draws: 0, losses: 0, goalsFor: 0, goalsAgainst: 0, goalDifference: 0, points: 0 });
        }
        return standingsMap.get(teamName)!;
      };

      for (const match of matches) {
        let homeTeam = "";
        let awayTeam = "";
        let homeGoals = 0;
        let awayGoals = 0;

        if ("home_team" in match) {
          homeTeam = match.home_team;
          awayTeam = match.away_team;
          homeGoals = match.home_goal;
          awayGoals = match.away_goal;
        } else if ("home" in match) {
          homeTeam = match.home;
          awayTeam = match.away;
          homeGoals = match.home_goal;
          awayGoals = match.away_goal;
        } else {
          homeTeam = match.Equipe_mandante;
          awayTeam = match.Equipe_visitante;
          homeGoals = match.Gols_mandante;
          awayGoals = match.Gols_visitante;
        }

        const home = getOrCreate(homeTeam);
        const away = getOrCreate(awayTeam);

        home.matches++;
        away.matches++;
        home.goalsFor += homeGoals;
        home.goalsAgainst += awayGoals;
        away.goalsFor += awayGoals;
        away.goalsAgainst += homeGoals;

        if (homeGoals > awayGoals) {
          home.wins++;
          home.points += 3;
          away.losses++;
        } else if (awayGoals > homeGoals) {
          away.wins++;
          away.points += 3;
          home.losses++;
        } else {
          home.draws++;
          away.draws++;
          home.points += 1;
          away.points += 1;
        }
      }

      const finalStandings = Array.from(standingsMap.values()).map((s) => ({
        ...s,
        goalDifference: s.goalsFor - s.goalsAgainst,
      }));

      finalStandings.sort((a, b) => {
        if (b.points !== a.points) return b.points - a.points;
        if (b.goalDifference !== a.goalDifference) return b.goalDifference - a.goalDifference;
        return b.goalsFor - a.goalsFor;
      });

      return {
        content: [
          {
            type: "text",
            text: JSON.stringify({ competition, season, standings: finalStandings }, null, 2),
          },
        ],
      };
    }
  );
}
